### PR TITLE
Fix race in log

### PR DESCRIPTION
**Pull request by zeripath**
A race has been detected in #1441 relating to getting log levels.

This PR protects the GetLevel and GetStacktraceLevel calls with a mutex.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by lunny** [comment]
To resolve the data race, the lock has to still be added into write loglevel in `internalResetLevel`.

**Comment by zeripath** [comment]
Both calls to that are already locked:

https://github.com/go-gitea/gitea/blob/a7fe58b84574dffdb2a0bcdbce32d661225bbc8d/modules/log/event.go#L205

https://github.com/go-gitea/gitea/blob/a7fe58b84574dffdb2a0bcdbce32d661225bbc8d/modules/log/event.go#L393

**Comment by lunny** [comment]
The data race problem is one goroutine read `MultiChannelledLog.level` but another write `MultiChannelledLog.level`. So I think you should add lock on https://github.com/go-gitea/gitea/blob/a7fe58b84574dffdb2a0bcdbce32d661225bbc8d/modules/log/event.go#L375

**Comment by zeripath** [comment]
It's already locked.

https://github.com/go-gitea/gitea/blob/a7fe58b84574dffdb2a0bcdbce32d661225bbc8d/modules/log/event.go#L197-L210

https://github.com/go-gitea/gitea/blob/a7fe58b84574dffdb2a0bcdbce32d661225bbc8d/modules/log/event.go#L390-L394

**Review by lunny — APPROVED** [review]

File: modules/log/event.go [hunk 1]
```diff
@@ -359,11 +359,15 @@ func (m *MultiChannelledLog) Flush() {
 
 // GetLevel gets the level of this MultiChannelledLog
 func (m *MultiChannelledLog) GetLevel() Level {
+	m.mutex.Lock()
```
Comment (lafriks) [inline modules/log/event.go]: imho `mutex` should be `RWMutex` and Get methods use `RLock` and `RUnlock`

Comment (zeripath) [reply modules/log/event.go]: The RWmutex requires two calls to lock and unlock and would be slower.

This is simply a lock around a single value. If performance were really an issue an `atomic.Load*` `atomic.Store*` might be quicker but a RWMutex isn't going to improve this.

**Comment by zeripath** [comment]
OK simple testing suggests that a RWMutex would be faster in the case of a lot of calls to GetLevel.

File: modules/log/event.go (hunk 1 above)
Comment (lafriks) [reply modules/log/event.go]: What do you mean two calls? It requires same number of calls but state read only calls is not locking each other

**Review by lafriks — APPROVED** [review]
There was no need to rename variable but otherwise lgtm

**Comment by zeripath** [comment]
Looks like this might be exposing another race in the test.

**Comment by zeripath** [comment]
Hmm... I think there might be no non-racey way to assert the batch_length so I'm going to remove that.

**Comment by zeripath** [comment]
make lgtm work

**Pull request merged**